c3_ProofOfWork/proofOfWork.js

```diff
@@ -0,0 +1,329 @@
+      sig: EthCrypto.sign(this.wallet.privateKey, getHash(unsignedTx))
+    }
+    //broadcast new tx to network
+    this.network.broadcast(this.pid, tx)
+  }
+
+  applyTransaction (tx) {
+    if (tx.contents.type === 'send') { // Send coins
+      // Check the from address matches the signature
+      const signer = EthCrypto.recover(tx.sig, getHash(tx.contents))
+      if (signer !== tx.contents.from) {
+        throw new Error('Invalid signature!')
+      }
+      // Check there is enough balance for tx
+      if (this.state[[tx.contents.from]].balance - tx.contents.amount < 0) {
+        console.log('not enough money...yet')
+        return
+      }
+      // Check that the nonce is correct for replay protection
+      if (tx.contents.nonce !== this.state[[tx.contents.from]].nonce) {
+        if (!(tx.contents.from in this.invalidNonceTxs)) {
+          this.invalidNonceTxs[tx.contents.from] = {}
+        }
+        this.invalidNonceTxs[tx.contents.from][tx.contents.nonce] = tx
+        return
+      }
+      // If we don't have a record for this address, create one
+      if (!(tx.contents.to in this.state)) {
+        this.state[[tx.contents.to]] = {
+          balance: 0,
+          nonce: 0
+        }
+      }
+      // Update state
+      this.state[[tx.contents.from]].balance -= tx.contents.amount
+      this.state[[tx.contents.to]].balance += tx.contents.amount
+      this.state[[tx.contents.from]].nonce += 1
+    } else if (tx.contents.type === 'mint') { //Block reward
+      // Check the from address matches the signature
+      const signer = EthCrypto.recover(tx.sig, getHash(tx.contents))
+      if (signer !== tx.contents.to) {
+        throw new Error('Invalid signature in mint tx!')
+      }
+      // Update state
+      this.state[[tx.contents.to]].balance += tx.contents.amount
+    } else {
+      throw new Error('Invalid transaction type!')
+    }
+  }
+
+  applyInvalidNonceTxs (address) {
+    const targetNonce = this.state[address].nonce
+    if (address in this.invalidNonceTxs && targetNonce in this.invalidNonceTxs[address]) {
+      this.applyTransaction(this.invalidNonceTxs[address][targetNonce])
+      delete this.invalidNonceTxs[address][targetNonce]
+      this.applyInvalidNonceTxs(address)
+    }
+  }
+
+  tick () {
+    /*
+     * In this sim we will not send random transactions
+     *
+
+    // If we have no money, don't do anything!
+    if (this.state[this.wallet.address].balance < 10) {
+      console.log('We are honest so we wont send anything :)')
+      return
+    }
+    // Generate random transaction
+    const tx = this.generateTx(nodes[Math.floor(Math.random() * nodes.length)].wallet.address, 10)
+    this.transactions.push(tx)
+    this.applyTransaction(tx)
+    // Broadcast this tx to the network
+    this.network.broadcast(this.pid, tx)
+    */
+  }
+
+  // Fork choice
+  // Only apply transactions which are contained in the longest chain
+  // and returns the resulting state object.
+  getState() {
```

Comment:
Am I right that if you receive a block which has a `prevhash` pointer to a block that you haven't received yet then function will break?

Reply:
Yes that's correct. I will review and submit a new PR. 

Thanks.